Visual effects<新增生命值,新增出界效果>

YiKaiWu1125 opened the pull request:
## Types of changes


+  **New feature**

## Description
+ [x] 四個遊戲加入電管出界後的視覺效果
+ [x] 使'elbow'mode可順利運行
+ [x] 新增生命功能

## Related Issue

解決 #20

YiKaiWu1125 committed:
四個遊戲加入電管出界後的視覺效果

 + 在迷宮模式新增觸碰到管子的視覺效果(變色管)
 	+ remove_background有點怪怪的無法被套用，只能套用在舊版easy.png，其餘新圖片皆不能是為甚麼呀?
	+ esay.png 有更新過喔.
	+ 新增outpipe_easy.png與playing_easy.png.
	+ maze_game.py 新增 outtime 用來記錄需要亮'紅管'的剩餘時間.
 + 對cube新增 電管出界後一片紅
 + 對base_video_game新增 邊線重劃可解決輕微重疊問題
 + 對base_video_game新增 出界電管變紅

YiKaiWu1125 committed:
使'elbow'mode可順利運行

+ 在soultion.py 新增 elbow.

YiKaiWu1125 committed:
修正video與battle無傳送 outpipe 與 影片鏡頭顛倒.

+ 修正video與battle無傳送 outpipe .
+ 修正base_video_game播影片時會也會salc的問題，因而導致最一開始播完影片就會直接被判斷出界(已修復) .
+ 影片與攝像機左右顛倒問題(已修復) .

YiKaiWu1125 committed:
新增生命功能

+ 還需搭配前端監聽 heart_reset 來重製破裂愛心(僅brattle模式下會用到).

YiKaiWu1125 committed:
更換video模式影片

asas1asas200 commented on maze_game.py:
```diff
@@ -23,8 +23,12 @@ def remove_background(image):
+        self.outpipe_img = cv2.imread(
+            'static/images/outpipe_' + self.level + '.png', cv2.IMREAD_UNCHANGED)
+        self.playing_img = cv2.imread(
+            'static/images/playing_' + self.level + '.png', cv2.IMREAD_UNCHANGED)
```
`f'static/images/outpipe_{self.level}.png'`
`f'static/images/playing_{self.level}.png'`

YiKaiWu1125 replied:
+ [x] 已解決

asas1asas200 commented on solution.py:
```diff
@@ -43,7 +45,7 @@ def draw_landmarks(self, image, results):
                     self.mp_drawing_styles.get_default_hand_connections_style()
                 )
-        elif self.body == 'foot' or self.body == 'head':
+        elif self.body == 'foot' or self.body == 'head' or self.body == 'elbow':
```
Because there are many types to be compared, it can write as below:
```python
elif self.body in ['foot', 'head', 'elbow']:
```

YiKaiWu1125 replied:
+ [x] 已解決

asas1asas200 commented on maze_game.py:
```diff
@@ -55,34 +59,53 @@ def is_touched(x, y):
+        if self.life <= 0:
+            if self.status != 'game_over':
+                self.gameover = True
+                self.status = "game_over"
```
This function is named `draw()`, so it should only draw on the image, the logical operations should be put in `calc()`.

asas1asas200 commented on tube_game.py:
```diff
@@ -35,43 +37,65 @@ def calc(self, results):
+        if self.life <= 0:
+            if self.status != 'game_over':
+                self.gameover = True
+                self.status = "game_over"
```
Ditto.

asas1asas200 commented on tube_game.py:
```diff
@@ -35,43 +37,65 @@ def calc(self, results):
         image = cv2.resize(image, (1280, 720))
         return cv2.imencode('.jpg', image)[1].tobytes()
+
+    def draw_outpipe(self,image):
+        image = cv2.addWeighted(image, 0.4, self.outpipe_img, 0.6, 0)
+        return image
```
You don't need to write a new function for doing this.

YiKaiWu1125 replied:
我以為這樣可以使程式可視度高一點的說.

asas1asas200 replied:
> 我以為這樣可以使程式可視度高一點的說.

Just add a comment at line end like other cv2 drawer function.

asas1asas200 commented on video_game.py:
```diff
@@ -34,5 +36,8 @@ def calc(self, results):
         image = super().video_draw(
             results, image) if self.reading else super().draw(results, image)
+        if self.life <= 0:
+                image = cv2.putText(image, "you are lose", (0, 150),
```
`You lose`

asas1asas200 commented on app.py:
```diff
@@ -87,7 +87,9 @@ def gen(camera):
         if camera.get_game().check_outpipe():
             socketio.emit('out_pipe', {'data': 'out_pipe'})
-            print("out_pipe")
+        if camera.get_game().check_heart_reset():
+            socketio.emit('heart_reset', {'data': 'heart_reset'})
+            print("heart reset.")
```
Only `battle game` owns this variable.
This function may not need, it can do in frontend.
Check the heart number while websocket received, then redraw the heart.
TBD.

YiKaiWu1125 replied:
此題無解~~

asas1asas200 replied:
Add a FIXME comment for this problem then merge this PR.

YiKaiWu1125 committed:
細部修改部分bug

+ maze_game.py 簡化引入圖片鏈結.
+ maze_game.py 簡化判斷式.
+ maze_game.py 移動if self.status != 'game_over': 從draw()到calc().
+ tube_game.py 移動if self.status != 'game_over': 從draw()到calc().
+ tube_game.py 移除draw_outpipe function.
+ video_game.py 修正顯示字串 'you are lose' -> 'you lose'.
